app/models/terminal_executor.rb

```diff
@@ -131,7 +134,18 @@ def script(commands)
   end
 
   def print_and_execute(c, resolve: true)
-    "echo » #{c.shellescape}\n#{resolve ? resolve_secrets(c) : c}"
+    if c.include? HIDDEN_PREFIX
+      print_and_execute_hidden_command(c)
+    else
+      "echo » #{c.shellescape}\n#{resolve ? resolve_secrets(c) : c}"
+    end
+  end
+
+  def print_and_execute_hidden_command(c)
+    # Replace `hidden://abcd` up until the first doublequote with HIDDEN_TXT
+    hidden_command   = c.sub(/#{HIDDEN_PREFIX}[^"]+/, HIDDEN_TXT)
+    resolved_command = c.sub(/#{HIDDEN_PREFIX}/, '')
+    "echo » #{hidden_command.shellescape}\n#{resolved_command}"
   end
```

Comment:
- echo logic is duplicated in 2 places
 - hidden logic is checked in 2 spaces
 - initial hidden check might not match hidden replaement

```Ruby
def print_and_execute(c, resolve: true)
  print_and_execute_hidden_command(c) ||
    print_and_execute_raw(c, resolve ? resolve_secrets(c) : c)
end

def print_and_execute_hidden_command(c)
  return unless hidden_command = c.dup.sub!(/#{HIDDEN_PREFIX}[^"]+/, HIDDEN_TXT)
  resolved_command = c.sub(HIDDEN_PREFIX, '')
  print_and_execute_raw hidden_command, resolved_command
end

def print_and_execute_raw(print_command, real_command)
  "echo » #{print_command.shellescape}\n#{real_command}"
end
```

Reply:
Good catch with the double hidden check, frozen strings led me down that route but the `dup.sub!` approach is really nice.